### ISSUE TITLE
`IntoDatum::is_compatible_with` should need `unsafe`

### ISSUE THREAD

**Issue by workingjubilee** [pgcentralfoundation/pgrx#1361]
Likely `unsafe trait` for an `unsafe impl`, because we rely on its implementation being correct. :grimacing:

We already know `IntoDatum` should be an `unsafe trait`, but I'm marking it in a separate issue because we should probably factor it out from `IntoDatum`.

**Comment by workingjubilee** [comment]
@eeeebbbbrrrr I know that Postgres does it, and to _some_ extent that's certainly "motivation enough", but what was the motivation for having "binary compatibility" be a factor when extracting/transferring types on the Rust side? I'm especially thinking of how it's used in our heap tuple code.

**Comment by eeeebbbbrrrr** [comment]
Yeah.  Because the type compatibility can't be figured out at compile time.

For example, "VARCHAR" and "TEXT" types are binary compatible but have different type oids.  pgrx maps the Rust `String` and `&str` types to `TEXTOID`, but then we'd fail at runtime with a type conversion error if the datum type was actually `VARCHAROID`.

I don't know that there's a lot of other examples of this, but there likely are.  I can imagine users wanting two UDTs with the same structure to be interchangeable as well.  Those would have different type oids but be "the same" bitwise.

**Comment by workingjubilee** [comment]
@eeeebbbbrrrr if the relationship is many-to-many, then why are we enforcing a 1:1 relationship on the datum end? can't this surface bugs in different ways, where we decode something from a type given as VARCHAROID and then return it by asserting it is TEXTOID?

**Comment by eeeebbbbrrrr** [comment]
> where we decode something from a type given as VARCHAROID and then return it by asserting it is TEXTOID?

That's the point of "is binary compatible".  At the stage where we're validating the Datum can be used with the desired Rust `T`, there's no difference.

What am I missing?

**Comment by workingjubilee** [comment]
Some off-sides discussion happened: the main "thing" here that perturbs me is that we effectively have a serialization framework and not all our choices are exactly sensical, or if they are, they aren't _systematic_.

However we do have to have _some_ version of this in order to satisfy the oddities of e.g. VARCHAR and TEXT being actually-the-same-thing.
